### ISSUE TITLE
External storage and SD card

### ISSUE THREAD

**Issue by Lingomat** [langtech/aikuma#302]
The Sony Xperia M appears to have two external storage mount paths. One based on internal flash and the second being an SD card. Aikuma presently chooses the first one. This is limited to 2GB in size with barely 1GB free on this phone. While this isn't going to cause immediate problems it's obviously desirable that the larger true SD mount is chosen as the file storage path. It's probably also desirable to choose SD storage since that can facilitate fast local transfers, e.g. mounting as a USB mass storage device, or physically moving the SD card.

@lisaslyis knows more since he's seen the phone.

**Comment by stevenbird** [comment]
Information about the mount points is stored in a system file, and not something available via the Android API. So we cannot confidently distinguish the mount points from the system file. Perhaps we can inspect the top-level directory names in each mount? This seems too messy. Can we present available options to the user? (But what information can we give the user so s/he can decide?)

**Comment by stevenbird** [comment]
Let's look at what other media apps do, and whether they allow a user to manually select the external SD card over the internal storage.

**Comment by stevenbird** [comment]
Check a phone which has an micro-SD card slot to see if there are two mount points (in Android File Transfer). Try moving the `/aikuma` folder between mount points. Can Aikuma still find it?

(This means a user could manually move the `/aikuma` folder to the biggest mount.)

**Comment by lisaslyis** [comment]
In case of Sony Xperia-E1, micro-SD card and internal storage are recognized as separate two mount points in the laptop file explorer.

**Comment by stevenbird** [comment]
Suppose that Aikuma searched all mount points for a `/aikuma` folder, and reported an error if it found more than one. Then a user could move the folder between mounts (presumably to the larger one).

**Comment by stevenbird** [comment]
However, in order to identify the mount points, we need to inspect a system file directly; this isn't supported by Android. For now, the advice is to use phones with substantial built-in storage.

**Comment by Lingomat** [comment]
Android's support of SD cards has been patchy and under flux. As of 4.4 the situation is much clearer.

When you call Context.getExternalFilesDirs(), the first result is primary storage, any others are 'secondary storage' which is where SD cards will end up. All the usual flags apply, like testing for whether it's removable or not.

Analysis here: http://possiblemobile.com/2014/03/android-external-storage/

**Comment by stevenbird** [comment]
Thanks @Lingomat. This suggests that it would be possible to search all mount points for an `/aikuma` folder (as mentioned earlier). I'm happy to require Android 4.4 if it resolves this issue.

**Comment by lisaslyis** [comment]
As I mentioned in the meeting, Android allows read-permission but not write-permission. So moving files from the app-interface will not be possible.

This is also mentioned In the link: http://possiblemobile.com/2014/03/android-external-storage/
Secondary volumes are fully readable with the global read permission granted.  If a user were to dump a bunch of files in random locations on an SD card and insert it into the device, any application could still read them.  Third-party applications just can’t add more files of their own in random locations.

**Comment by Lingomat** [comment]
I'm not sure that's true. This text is discussing the ability to write files in any location on external secondary volumes. We don't want to do that. The point is this: If there is a an external secondary volume and this is an SD card, I believe we can use the handle that Android provides to gain app-specific write permission on this volume.

Which is what we want, right? We don't necessarily want to adopt an old Aikuma folder on SD and then start writing to it. We want to use that volume to save files, by letting the OS choose the path on that volume, the only reason we care is because that volume has space and the internal one doesn't.

Also, at the end of that article, the author suggests that Android is moving towards a model with rules such that external secondary becomes a default location for large files. That's interesting.

**Comment by lisaslyis** [comment]
@Lingomat. I might have been unclear about the use case of SD card. I thought that aikuma-folder-architecture needs to be created at the root of SD card.
As an aside, even this(write-access to the root of SD card) might be possible from Android 5. I'm not sure about this yet. But it seems like Android tries to support a full access to SD card from version 5.

**Comment by stevenbird** [comment]
I think it is premature to resolve this issue, unfortunately. We may need to wait until 5.0 is more widespread.

**Comment by Lingomat** [comment]
Okay. Of course we can somewhat alleviate this problem by using compression. We've discussed this previously. I'm not sure what the objection to aac was, since that's built-in. Something about accurate seeking?

Shame Android only decodes FLAC. There is a Java library for this though:
http://jflac.sourceforge.net/

**Comment by stevenbird** [comment]
I'm putting this in the too-hard basket for now.

**Comment by Lingomat** [comment]
Okay. What this does mean, though, is that I have to buy phones that have about 16gb of storage. This removes the majority of dirt cheap commodity devices. I'll end up having to pay nearer $200 a handset rather than $100. Well, actually, I might buy some refurbed brand name models which I can get for about US$115 a shot, if I find one with a reasonable build of Android.

**Comment by LaurenGawne** [comment]
It also removes a potential backdoor for removing recordings from devices,
or copying files across in the absence of a functioning network.

**Comment by Lingomat** [comment]
I take it you mean a desirable backdoor?

**Comment by LaurenGawne** [comment]
um, yeah - I find it useful in Nepal when using small numbers of phones in
places where I can't get a network happening

**Comment by Lingomat** [comment]
Yeah I thought so, it's just I haven't normally heard backdoor in a positive sense. :) We're definitely on the same page anyway. Unfortunately this really is quite a hard thing, as discussed earlier. I'm also keen to get the ball rolling with video but it's not clear to me if I'll use the video camera on phones or if I'll just do things myself with a video camera. 

I suspect this is something I'll work out in the field. An environment where Taipei is an hour up the road is quite helpful for techie fieldwork no? :)

**Comment by Lingomat** [comment]
The upcoming Android M looks like it will completely solve this issue. Users will be able to insert an SD card and format this to expand *internal* storage. This also means that upcoming devices (like Nexus phones etc) will be adding SD cards where previously they relied on internal memory only.

This also means that very cheap handsets with SD cards (which exist today) will be good platforms for Aikuma. No more having to buy high-end devices with 16GB storage!

http://www.androidpolice.com/2015/05/28/android-m-feature-spotlight-external-storage-can-be-adopted-as-true-internal-storage-or-accessed-normally-with-no-additional-apps/
